repo/repo.go

```diff
@@ -1,282 +1,331 @@
+func httpDefaultError(w http.ResponseWriter, code int) {
+	http.Error(w, http.StatusText(code), code)
+}
 
-	return false
+// httpMethodNotAllowed writes a 405 Method Not Allowed HTTP error with
+// the required Allow header listing the methods that are allowed.
+func httpMethodNotAllowed(w http.ResponseWriter, allowed []string) {
+	w.Header().Set("Allow", strings.Join(allowed, ", "))
+	httpDefaultError(w, http.StatusMethodNotAllowed)
 }
 
-// join takes a number of path names, sanitizes them, and returns them joined
-// with base for the current operating system to use (dirs separated by
-// filepath.Separator). The returned path is always either equal to base or a
-// subdir of base.
-func join(base string, names ...string) (string, error) {
-	clean := make([]string, 0, len(names)+1)
-	clean = append(clean, base)
-
-	// taken from net/http.Dir
-	for _, name := range names {
-		if !valid(name) {
-			return "", errors.New("invalid character in path")
-		}
+// BlobPathRE matches valid blob URI paths with optional object IDs
+var BlobPathRE = regexp.MustCompile(`^/(data|index|keys|locks|snapshots)/([0-9a-f]{64})?$`)
 
-		clean = append(clean, filepath.FromSlash(path.Clean("/"+name)))
-	}
+// ObjectTypes are subdirs that are used for object storage
+var ObjectTypes = []string{"data", "index", "keys", "locks", "snapshots"}
+
+// FileTypes are files stored directly under the repo direct that are accessible
+// through a request
+var FileTypes = []string{"config"}
 
-	return filepath.Join(clean...), nil
+func isHashed(objectType string) bool {
+	return objectType == "data"
 }
 
-// getRepo returns the repository location, relative to s.Path.
-func (s *Server) getRepo(r *http.Request) string {
-	if strings.HasPrefix(fmt.Sprintf("%s", middleware.Pattern(r.Context())), "/:repo") {
-		return pat.Param(r, "repo")
-	}
+// BlobOperation describe the current blob operation in the BlobMetricFunc callback
+type BlobOperation byte
 
-	return "."
-}
+const (
+	BlobRead   = 'R' // A blob has been read
+	BlobWrite  = 'W' // A blob has been written
+	BlobDelete = 'D' // A blob has been deleted
+)
 
-// getPath returns the path for a file type in the repo.
-func (s *Server) getPath(r *http.Request, fileType string) (string, error) {
-	if !s.isValidType(fileType) {
-		return "", errors.New("invalid file type")
+// BlobMetricFunc is the callback signature for blob metrics. Such a callback
+// can be passed in the Options to keep track of various metrics.
+// objectType: one of ObjectTypes
+// operation: one of the BlobOperations above
+// nBytes: the number of bytes affected, or 0 if not relevant
+type BlobMetricFunc func(objectType string, operation BlobOperation, nBytes uint64)
+
+// ServeHTTP performs strict matching on the repo part of the URL path and
+// dispatches the request to the appropriate handler.
+func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	urlPath := r.URL.Path
```

Comment:
I think we should ensure that the URL path here is absolute and all references to higher directories (`..`) are resolved. Usually, this can be accomplished as follows:
```
urlPath := path.Clean(path.Join('/'+r.URL.Path))
```

Reply:
The path validation is done in a different place, and the Repo package performs very strict path matching to ensure safety.

Adding this `path.Clean` would actually make it *less* secure, because you would have multiple urls that point at the same resource.

Consider a situation where a sysadmin adds an HTTP proxy in front of the service so that `/metrics` is not accessible externally. With the Clean code, a path like `/some-username/../metrics` could return the metrics, while passing the proxy in front of it. With the current validation code this attack would not work.

Reply:
Understood, thanks! I only read `repo.go` so far, and I even saw the strict matching. :)